File: parent/518 -- Coin Change II/tutorial2.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        # Bottom-Up DP (most efficient. improves Space)
        # Time: O(amount*n)
        # Space: O(amount)

        dp = [0] * (amount+1)
        dp[0] = 1

        for c in coins:
            for i in range(c, amount+1):
                if i >= 0:
                    dp[i] += dp[i-c]

        return dp[-1]
            

File: parent/518 -- Coin Change II/tutorial2.go
package main

func change(amount int, coins []int) int {
	dp := make([]int, amount+1)
	dp[0] = 1

	for _, c := range coins {
		for i := c; i <= amount; i++ {
			if i-c >= 0 {
				dp[i] += dp[i-c]
			}
		}
	}

	return dp[amount]
}
